Update MIPI DSI driver to support ULPS

KATE-WANG-NXP opened the pull request:
Add new item ulps_control in binding. If the MIPI DSI on the SoC support ULPS, and user set the bus to enter ULPS after transfer in mipi_dsi_msg, driver will set the bus to enter ULPS.

KATE-WANG-NXP commented on drivers/mipi_dsi/dsi_mcux_2l.c:
```diff
@@ -146,9 +188,12 @@ static int dsi_mcux_tx_color(const struct device *dev, uint8_t channel,
 
 	k_sem_give(&data->transfer_sem);
+
+	dsi_mcux_transfer_complete(dev);
```
Yes, fixed

KATE-WANG-NXP commented on drivers/mipi_dsi/dsi_mcux_2l.c:
```diff
@@ -353,6 +410,28 @@ static int dsi_mcux_attach(const struct device *dev,
 						dsi_pixel_clk_freq, dphy_bit_clk_freq);
 	}
 
+	/* Calculate the delay before entering ULPS. */
```
Added

KATE-WANG-NXP commented:
Hi all, please help review this PR, thanks

ngphibang commented on drivers/mipi_dsi/dsi_mcux_2l.c:
```diff
@@ -392,6 +494,18 @@ static ssize_t dsi_mcux_transfer(const struct device *dev, uint8_t channel,
+	data->flags |= (msg->flags & MIPI_DSI_MSG_USE_LPM) ? MIPI_DSI_MSG_USE_LPM : 0U;
+	if (msg->flags & MCUX_DSI_2L_ULPS) {
+		data->flags |= MCUX_DSI_2L_ULPS;
+		if (!DT_PROP(DT_NODELABEL(mipi_dsi), ulps_control)) {
```
You will have macro expansion error if `ulps_control` property does not exist.

KATE-WANG-NXP replied:
Fixed

ngphibang replied:
The line `if (!DT_PROP(DT_NODELABEL(mipi_dsi), ulps_control)) {` still there

ngphibang commented on include/zephyr/drivers/mipi_dsi/mipi_dsi_mcux_2l.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2023 NXP
+ * Copyright 2023,2024 NXP
```
2025

ngphibang commented on drivers/mipi_dsi/dsi_mcux_2l.c:
```diff
@@ -392,6 +494,20 @@ static ssize_t dsi_mcux_transfer(const struct device *dev, uint8_t channel,
 	dsi_xfer.flags = (msg->flags & MIPI_DSI_MSG_USE_LPM) ? 0 : kDSI_TransferUseHighSpeed;
 
+#if DT_PROP(DT_NODELABEL(mipi_dsi), ulps_control)
+	data->flags = 0U;
```
I think we don't need to set it to 0 first to make an OR later.

ngphibang commented on drivers/mipi_dsi/dsi_mcux_2l.c:
```diff
@@ -392,6 +494,20 @@ static ssize_t dsi_mcux_transfer(const struct device *dev, uint8_t channel,
 
+#if DT_PROP(DT_NODELABEL(mipi_dsi), ulps_control)
+	data->flags = 0U;
+	data->flags |= (msg->flags & MIPI_DSI_MSG_USE_LPM) ? MIPI_DSI_MSG_USE_LPM : 0U;
```
`data->flags = (msg->flags & MIPI_DSI_MSG_USE_LPM) ? MIPI_DSI_MSG_USE_LPM : 0U;`

ngphibang commented:
Please fix CI errors

KATE-WANG-NXP committed:
drivers: mipi_dsi: dsi_mcux_2l: Update driver to support ULPS

Add new item ulps_control in binding. If the MIPI DSI on the SoC support
ULPS, and user set the bus to enter ULPS after transfer in mipi_dsi_msg,
driver will set the bus to enter ULPS.

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>

ngphibang reviewed (approved):
Overall LGTM

KATE-WANG-NXP commented:
Hi all, please help review this PR, the v4.1.0(Feb14) is quite close, thanks
